### PR TITLE
Use security token when making Monit API calls

### PR DESCRIPTION
**Pull request by cdonati**
The cookie field was introduced in Monit 5.20.

This also reduces the number of Monit CLI calls. Monit was recently patched in Ubuntu to address https://security-tracker.debian.org/tracker/CVE-2016-7067. However, the CLI client was not modified to match the new restrictions on the HTTP interface.

https://bugs.launchpad.net/ubuntu/+source/monit/+bug/1786910

These changes are not exhaustive. There are probably more invocations of the monit script that are ineffectual. This just changes enough to pass most of the tests.

### CODE REVIEW

**Comment by cdonati** [comment]
https://ocd.appscale.com:8080/job/Daily%20Build/5319/

File: common/appscale/common/monit_interface.py [hunk 1]
```diff
@@ -248,14 +257,74 @@ def send_command(self, process_name, command):
...
-      yield self._async_client.fetch(process_url, method='POST', body=payload)
+      yield self._async_client.fetch(
+        process_url, method='POST', body=urllib.urlencode(params),
+        headers=headers)
     except HTTPError as error:
-      if error.code == 404:
+      if error.code == httplib.FORBIDDEN:
+        # Retrieve CSRF token (introduced in Monit 5.20).
+        response = yield self._async_client.fetch(process_url)
+        self._csrf_token = re.search('securitytoken=([a-zA-Z0-9]+);',
```
Comment (scragraham) [inline common/appscale/common/monit_interface.py]: Not sure if we'll hit it with monit, but the Cookie header syntax doesn't require a semicolon, so its possible (probably unlikely) to see:
```
Cookie: securitytoken=12345
```
too bad the client doesn't have support for cookie handling/parsing. oh well.

File: AppManager/app_manager_server.py [hunk 2]
```diff
@@ -501,7 +478,9 @@ def unmonitor_and_terminate(watch):
     watch: A string specifying the Monit entry.
   """
   try:
-    unmonitor(watch)
+    monit_retry = retry(max_retries=5, retry_on_exception=MonitUnavailable)
```
Comment (jeanleonov) [inline AppManager/app_manager_server.py]: Default backoff sequence is 0.2, 0.4, 0.8, 1.6, 3.2, 6.4, 12.8,... If you set `max_retries=5` sum of pauses between retries will be `6.2s`. Is this usually enough for monit to become available again?
`retrying_timeout=30` or 60 could be an alternative.

**Comment by jeanleonov** [comment]
I suspect that failures in previous build was not transient. Trying one more time:
https://ocd.appscale.com:8080/job/Daily%20Build/5322/

File: common/appscale/common/monit_interface.py (hunk 1 above)
Comment (cdonati) [reply common/appscale/common/monit_interface.py]: Good point. Using Python's cookie library is more robust.

File: common/appscale/common/monit_interface.py [hunk 3]
```diff
@@ -408,3 +410,15 @@ def _reload(self):
...
+  def _parse_security_token(http_response):
+    """ Extracts the security token from an HTTP response.
+
+    Args:
+      A tornado.httpclient.HTTPResponse object.
+    Returns:
+      A string containing the security token.
+    """
+    cookie = SimpleCookie(http_response['Set-Cookie'])
+    return cookie['securitytoken'].value
```
Comment (scragraham) [inline common/appscale/common/monit_interface.py]: KeyError can be thrown here, if there are no cookies in the response, and if securitytoken isn't a cookie. I think the semantics should be that we return None if the token isn't found:
```
try:
    cookie = SimpleCookie(http_response['Set-Cookie'])
    return cookie['securitytoken'].value
except KeyError:
    return None
```

**Comment by cdonati** [comment]
Build for df4df1e:
https://ocd.appscale.com:8080/job/Daily%20Build/5324/
https://ocd.appscale.com:8080/job/Datastore/53/

**Comment by cdonati** [comment]
https://ocd.appscale.com:8080/job/Daily%20Build/5326/

**Pull request merged**